hal/common/mbed_wait_api_rtos.cpp

```diff
@@ -0,0 +1,56 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2013 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This implementation of the wait functions will be compiled only
+// if the RTOS is present.
+#ifdef MBED_CONF_RTOS_PRESENT
+
+#include "wait_api.h"
+#include "us_ticker_api.h"
+#include "rtos.h"
+#include "critical.h"
+
+void wait(float s) {
+    wait_us(s * 1000000.0f);
+}
+
+void wait_ms(int ms) {
+    wait_us(ms * 1000);
+}
+
+// Wait for the given number of microseconds using a hardware timer
+// in a busy wait loop
+static void wait_us_busy(int us) {
+    if (us > 0) {
+        uint32_t start = us_ticker_read();
+        while ((us_ticker_read() - start) < (uint32_t)us);
+    }
+}
+
+void wait_us(int us) {
+    // Use the RTOS to wait for millisecond delays if possible
+    int ms = us / 1000;
```

Comment:
Guessing there is a bit twiddling shift that is more appropriate to at least have a deterministic execution cycle count. Not to handle now but on M0 this will be a bit off but maybe it always has been.